### DISCUSSION TITLE
project proposal: maplibre-font-maker

### DISCUSSION THREAD

**Discussion by bdon** [maplibre/maplibre#144] [Ideas]
I'm proposing to move ~https://github.com/protomaps/sdf-glyph-tool~ https://github.com/maplibre/font-maker into the MapLibre GitHub org

This is, as far as I know, the only browser-based way to bulk-convert fonts into SDF files. It has a simple page at https://maplibre.org/font-maker/ that lets you upload one font file, click a button, and download a ZIP with all the SDF font range files. 

It does use the original https://github.com/mapbox/sdf-glyph-foundry library, which has not been updated in years, but the format itself is stable. It includes a command line utility with some more features. It's compiled to a WASM blob using Emscripten.

* I think this project belongs in the MapLibre org because:
  * it has little relevance outside of use within MapBox GL / MapLibre GL
  * Allowing users to bring their own fonts is an essential part of custom map frontend rendering
  * The MapLibre audience has many web developers, creating font files shouldn't require installing tools
 
* Arguments against including this in the MapLibre org:
  * There are other tools and approaches to do this, though most of them are command line only
  * The project is not feature-complete. It doesn't support multiple fonts in one fontstack, or Unicode locale variants yet. The web UI doesn't support combining multiple ranges of the same font into one SDF font, though the command line utility does. 

Based on discussion at https://github.com/maplibre/maplibre/discussions/132 I would suggest this exists at the lowest level of unsupported/hosted project. I would continue as the project maintainer. 

(EDIT: changed name from sdf-glyph-tool to font-maker)
(EDIT 2: repo has been migrated to https://github.com/maplibre/font-maker )

**Comment by HarelM** [comment]
I think it's a good idea.
There's a checklist to fill in order to have the repo under maplibre umbrella.

**Reply by HarelM** [reply]
If you open a new issue, you'll see the template there where you can check the tick boxes of the things you have completed in order to make the repo part of the maplibre projects.

**Reply by wipfli** [reply]
I think it's a good idea as well

**Reply by wipfli** [reply]
I would name it `maplibre-font-tool` because `maplibre-fonts` made me think that you are proposing that MapLibre develops some fonts or so...

**Reply by lseelenbinder** [reply]
I agree on the name change. I thought the same thing at first.

**Reply by bdon** [reply]
We could also choose to include as part of the repository some pre-compiled PBF ZIPs or raw files for popular open source fonts like DejaVu Sans, etc. That might make the name `maplibre-fonts` more suggestive as a place to get fonts, but I like the name `maplibre-font-tool` as well (or maybe `maplibre-font-maker` ?)

**Reply by wipfli** [reply]
Maker is better

**Reply by HarelM** [reply]
We have the demo tiles repo for hosting some fonts I think, I also think it's a good place for them. Storing the fonts is like storing compilation results 😀

**Reply by bdon** [reply]
Reflecting on the other thread about missing features, an argument against supplying many pre-rendered fonts in demotiles is that users might want to create multi-face fontstacks; we don't want to support all permutations of combined popular fonts, so instead `maplibre-font-maker` should be able to generate a fontstack on-demand from those preselected raw files. 

In practice, doing this by downloading TTFs over the web can consume an impractical amount of bandwidth. But we could add other constraints, like excluding CJK codepoint ranges which aren't needed with localIdeographs enabled.

I believe the other solution is to run a server to combine fonts into fontstacks on-the-fly, but that seems outside the scope of maplibre.

**Comment by lseelenbinder** [comment]
Would our [pbf_font_tools](https://github.com/stadiamaps/pbf_font_tools) be helpful for this @bdon? We support some of the missing functionality, but it's not JS ready.

**Reply by bdon** [reply]
It looks like pbf_font_tools supports creating a fontstack with multiple fonts e.g. `Font1,Font2` - this shouldn't be too much work to implement in sdf-glyph-tool right now, the time consuming part would building the Web UI for this. Open to PRs :)

I do believe the project as-is is useful for when each `text-font` rule only contains a single face. Maybe that's a minority of cases: if there's a strong demand among this audience for multi-face fontstacks, then that should be a priority.

Other features I named above are more niche. I don't think there's much use for Unicode locale variants in practice; the main affected script is Han which is semi-solved by the localIdeographs code path. I don't know about native, however.

**Comment by klokan** [comment]
Good idea. Just do not use word "maplibre" in the name of the tool.

**Reply by bdon** [reply]
we can use https://github.com/maplibre/font-maker as the repo

**Comment by bdon** [comment]
A couple updates:

* I have added the main missing feature: combining multiple OTF and TTF files into a single font stack.
* There's now a live preview using demotiles for viewing the font on the map (with test strings + sizes)
* The fontstack name is derived automatically from the FreeType data.
* Big UI cleanup.

<img width="1052" alt="Screen Shot 2022-12-03 at 1 49 41 PM" src="https://user-images.githubusercontent.com/77501/205426556-0d7ca726-4c9d-4541-9fff-ae8ad08775b4.png">

**Reply by HarelM** [reply]
Nice!!
